blockmanager.go

```diff
@@ -1019,109 +1022,153 @@ func (b *blockManager) getCheckpointedCFHeaders(checkpoints []*chainhash.Hash,
-			// them.
-			if startHeight > curHeight+1 {
-				log.Debugf("Got response for headers at "+
-					"height=%v, only at height=%v, stashing",
-					startHeight, curHeight)
+		msgs = append(msgs, &query.Query{
+			Req:        m,
+			HandleResp: handleResp,
+		})
+	}
 
-				queryResponses[checkPointIndex] = r
+	// Hand the queries to the work manager, and consume the verified
+	// responses as they come back.
+	errChan := b.cfg.QueryAccess.Query(msgs)
 
-				return true
-			}
+	// Keep waiting for more headers as long as we haven't received an
+	// answer for our last checkpoint, and no error is encountered.
+	for {
+		var r *wire.MsgCFHeaders
+		select {
+		case r = <-headerChan:
+		case err := <-errChan:
+			log.Errorf("Query finished with error (%v) before "+
+				"all responses received", err)
+			return
+		case <-b.quit:
+			return
+		}
 
-			// If this is out of order stuff that's already been
-			// written, we can ignore it.
-			if lastHeight <= curHeight {
-				log.Debugf("Received out of order reply "+
-					"end_height=%v, already written", lastHeight)
-				return true
-			}
+		checkPointIndex, ok := stopHashes[r.StopHash]
+		if !ok {
+			// We never requested a matching stop hash.
+			log.Errorf("did not request stopHash %v", r.StopHash)
+			return
+		}
 
-			// If this is the very first range we've requested, we
-			// may already have a portion of the headers written to
-			// disk.
-			//
-			// TODO(roasbeef): can eventually special case handle
-			// this at the top
-			if bytes.Equal(curHeader[:], initialFilterHeader[:]) {
-				// So we'll set the prev header to our best
-				// known header, and seek within the header
-				// range a bit so we don't write any duplicate
-				// headers.
-				r.PrevFilterHeader = *curHeader
-				offset := curHeight + 1 - startHeight
-				r.FilterHashes = r.FilterHashes[offset:]
-
-				log.Debugf("Using offset %d for initial "+
-					"filter header range (new prev_hash=%v)",
-					offset, r.PrevFilterHeader)
-			}
+		// Find the first and last height for the blocks
+		// represented by this message.
+		nextCheckPointIndex := checkPointIndex + maxCFCheckptsPerQuery - 1
+		if nextCheckPointIndex >= uint32(len(checkpoints)) {
+			nextCheckPointIndex = uint32(len(checkpoints)) - 1
+		}
 
-			curHeader, err = b.writeCFHeadersMsg(r, store)
-			if err != nil {
-				panic(fmt.Sprintf("couldn't write cfheaders "+
-					"msg: %v", err))
-			}
+		startHeight := checkPointIndex*wire.CFCheckptInterval + 1
+		lastHeight := (nextCheckPointIndex + 1) * wire.CFCheckptInterval
 
-			// Then, we cycle through any cached messages, adding
-			// them to the batch and deleting them from the cache.
-			for {
-				// Determine the next checkpoint index we should
-				// process.
-				checkPointIndex += maxCFCheckptsPerQuery
-				if checkPointIndex == uint32(len(checkpoints)) {
-					checkPointIndex = uint32(len(checkpoints)) - 1
-				}
+		log.Debugf("Got cfheaders from height=%v to "+
```

Comment:
Nit: could extend some of these lines to use the new space available.

Reply:
done after method extraction